### ISSUE TITLE
Migrate to official node-redis

### ISSUE THREAD

**Issue by nodegin** [taskforcesh/bullmq#2490]
**Is your feature request related to a problem? Please describe.**

ioredis is being non-actively maintained for a while, it seems will be deprecated soon.
https://github.com/redis/node-redis/issues/2658#issuecomment-1968104973

**Describe the solution you'd like**
Migrate to https://github.com/redis/node-redis

**Describe alternatives you've considered**
n/a

**Additional context**
n/a

**Comment by manast** [comment]
Yeah, this is a quite a sad situation, it almost seems like Redislabs is being run by a chicken without head... most frustrating is the lack of communication with the community, leaving everybody wondering what is going on.

But from BullMQ's point of view, there is not a single issue with IORedis that we are aware of, and would such an issue arise we will fork IORedis and fix it for us. The only feature that would be relevant in terms of performance would be to have support for RESPv3 but neither IORedis nor node-redis supports it.

The APIs of IORedis and node-redis are different enough to make the work of supporting node-redis non-trivial, and there is no guarantee it will be better maintained either, in fact the number of commits seems to be quite low too...

If somebody really wants to use node-redis, one possibility would be to create a wrapper on top of node-redis that implements the IORedis specific APIs that are used by BullMQ... but I do not think this is worthy right now.

**Comment by manast** [comment]
Btw, it could be that some security advisory arises in IORedis that forces us to fork it even though there are no other issues, but lets give Redislabs a chance to make things right.

**Comment by codeagencybe** [comment]
What about moving to keydb or dragonfly? Both are decent drop-in replacements for redis and both open source. 
in terms of performance, they both claim to be much faster than redis too. 
Sounds like multiple wins then?

**Comment by manast** [comment]
Dragonfly is supported, as it is 100% compatible with Redis (https://bullmq.io/news/101023/dragonfly-compatibility/). KeyDB did not properly work with BullMQ last time I tested and probably does still not work but I am not sure.

In any case this discussion is about the NodeJS client library and less about the change in Redis license. I do not think the new Redis license affects regular users, but instead hosting providers which will now need to get a commercial license agreement from Redis in order to provide the newer versions. If I have to guess, these changes will just encourage new forks and competition to fill the need for a robust in-memory database, so not bad news for current Redis users in the long run.

**Comment by ivan-kleshnin** [comment]
> Describe the solution you'd like
Migrate to https://github.com/redis/node-redis

It's not well supported either 😞 
Look at the number of issues, at the absence of comments there, hanging PRs, look at the release history.

**Comment by jamesarosen** [comment]
> But from BullMQ's point of view, there is not a single issue with IORedis that we are aware of, and would such an issue arise we will fork IORedis and fix it for us.

That's true, but there are plenty of bugs in ioredis that don't directly affect BullMQ. If a job uses Redis for something else (like geodist calculations), it likely needs to load two redis clients. That's a waste of developer time, CPU time, and CO2.

**Comment by elovin** [comment]
I doubt that the official redis client will address any issues related to one of the redis forks.
And since the usage of redis forks will increase since the redis license change I do not think it is a good idea to replace ioredis with node-redis.

**Comment by elovin** [comment]
Ups I totally missed that ioredis is also part of the redis org.

**Comment by lukas-becker0** [comment]
FYI @manast , the [valkey](https://github.com/valkey-io) (AFAIK this is now the Redis fork with the biggest budget, [since a number of big companies are backing it](https://www.linuxfoundation.org/press/valkey-welcomes-new-partners-amid-growing-momentum).) guys already forked ioredis and call it [iovalkey](https://github.com/valkey-io/iovalkey) now.

**Comment by manast** [comment]
As long as IORedis works reliably for us, I do not see any benefit in using any other client... it will just introduce more risk and more likely new issues until all edge cases are ironed out.

**Comment by sirrodgepodge** [comment]
https://github.com/redis/ioredis/issues/1865

^This one is busted :/, reason enough to switch?

**Comment by manast** [comment]
@sirrodgepodge the thing is, how do we know node-redis does not have other issues that would impact us? so if we do all the migration work (which is not completely trivial), and then we discover other break-dealer issues, what then?
I think the most sensible thing to do is try to fix the issues in ioredis that affect BullMQ, worst case we could fork the project, but that is not an ideal situation either. It is quite incredible that Redis took over the maintenance of IORedis just to not maintain it. At least when the maintainer was @luin it tried its best to solve the issues...

**Comment by lukas-becker0** [comment]
@sirrodgepodge, seems like the issue you linked is "ElastiCache Serverless" specific, since people using
regular redis or dragonfly do not seem to have that issues.

**Comment by ozonep** [comment]
Instead of creating yet another fork, perhaps you could consider adopting and contributing to the 'iovalkey' client (like mentioned above)? My team has had a very positive experience with it. We submitted a PR to the `iovalkey` repository with stability fixes for clusters (no new features), and it was reviewed and merged within a week. In contrast, our original PR on `ioredis` has been open since March without any comments from the maintainers and will probably never be reviewed and merged. Additionally, the `iovalkey` client currently doesn't introduce any breaking changes compared to `ioredis` — just bug fixes.

**Comment by avifenesh** [comment]
Hi, 
valkey-glide maintainer, and AWS elasticache team member.
In addition to https://github.com/taskforcesh/bullmq/issues/3083,
I offer valkey-glide as a replacer, and we are willing to help with the migration.
Glide offers the best implementation of PubSub with the best ability to resubscribe on reconnection.
We offer the best fault tolerance (please test us).
We have team members from AWS elasticache/memoryDB and GCP memory store involved deeply in the development, we are part of valkey org, and we have a large community (59 contributors).
I'm the Node “leader” and I will love to see glide being integrated and packages as yours.
So if it's relevant, I offer my support for the migration, as part of the planning, and implementation.
Please let me know WDYT.

**Comment by manast** [comment]
@avifenesh hello!
This is the first time I heard about GLIDE to be honest, but I checked documentation and it looks like an interesting project, where you have focused on reliability which is really important for BullMQ users as well. 
I have been thinking for a while to support other clients such as node-redis, however in order to do this, we need to refactor the internals of BullMQ to use a generic Redis client interface that can then be implemented with IORedis, Glide or any other. Its a bit of work actually but it is doable, once this generic interface is available it should be relatively straightforward to implement it for Glide.

**Comment by avifenesh** [comment]
ah, and we have a python client as well :)
Also, Java. 
Go just released to public preview, and C# is next.
CPP and ruby are under community development.
So you might be able to grow beyond :)

**Comment by dulismo** [comment]
I needed a node-redis client for passwordless authentication with Azure Managed Redis, so I wrote a small wrapper adding ioredis compatible interfaces to a node-redis client, and so far it seems to be working. I only implemented the methods required by BullMQ, so that may be fairly representative of the work required if we wanted to integrate node-redis within bullmq, but maybe I completely overlooked some parts ...

```js
const definedCommands = new Map();
export function nodeRedisToIoRedis(redis: RedisClientType) {
  function ioredisWrapper(ioredis: any) {
    ioredis.status = 'ready'; // Needed by bullmq RedisConnection waitUntilReady()
    ioredis.isCluster = false;

    definedCommands.forEach((handler, name) => {
      redis[name] = handler.bind(redis);
    });

    ioredis.client = () => {};

    // hmaps
    ioredis.hset = async function hset(
      key: string,
      field: string,
      value: string
    ) {
      const result = await ioredis.hSet(key, field, value);
      return result;
    };
    ioredis.hget = async function hget(key: string, field: string) {
      const result = await ioredis.hGet(key, field);
      return result;
    };
    ioredis.hgetall = async function hgetall(key: string) {
      const result = await ioredis.hGetAll(key);
      return result;
    };
    ioredis.hmset = async function hmset(key: string, values: object) {
      const result = Promise.all(
        Object.entries(values).map(([field, value]) =>
          ioredis.hSet(key, field, value)
        )
      );
      return await result;
    };

    ///
    ioredis.bzpopmin = async function bzPopMin(key: string, timeout: number) {
      const result = await ioredis.bzPopMin(key, timeout);
      return result ? [result.key, result.value, result.score] : null;
    };

    // Lists
    ioredis.lpos = async function lpos(key: string, value: string) {
      const result = await ioredis.lPos(key, value);
      return result;
    };
    ioredis.ltrim = async function ltrim(
      key: string,
      start: number,
      stop: number
    ) {
      const result = await ioredis.lTrim(key, start, stop);
      return result;
    };

    // Sorted sets
    ioredis.zadd = async function zadd(
      key: string,
      score: number,
      value: string
    ) {
      const result = await ioredis.zAdd(key, [
        {
          score,
          value,
        },
      ]);
      return result;
    };
    ioredis.zrange = async (
      key: string,
      start: number,
      end: number,
      ...opts: string[]
    ) => {
      if (opts.includes('WITHSCORES')) {
        const result = await redis.zRangeWithScores(key, start, end, {
          REV: false,
        });
        return result.map((cur) => [cur.value, cur.score]).flat();
      }
      return await redis.zRange(key, start, end, {
        REV: false,
      });
    };
    ioredis.zrevrange = async (
      key: string,
      start: number,
      end: number,
      ...opts: string[]
    ) => {
      if (opts.includes('WITHSCORES')) {
        const result = await redis.zRangeWithScores(key, start, end, {
          REV: true,
        });
        return result.map((cur) => [cur.value, cur.score]).flat();
      }
      return await redis.zRange(key, start, end, {
        REV: true,
      });
    };
    ioredis.zcard = async function zcard(key: string) {
      const result = await ioredis.zCard(key);
      return result;
    };
    ioredis.zscore = async function zscore(key: string, member: string) {
      const result = await ioredis.zScore(key, member);
      return result;
    };

    // Streams
    ioredis.xadd = async function xadd(...args) {
      const result = await ioredis.sendCommand(
        ['XADD', ...args].map((cur) =>
          typeof cur === 'number' ? `${cur}` : cur
        )
      );
      return result;
    };
    ioredis.xtrim = async function xtrim(...args) {
      const result = await ioredis.sendCommand(
        ['XTRIM', ...args].map((cur) =>
          typeof cur === 'number' ? `${cur}` : cur
        )
      );
      return result;
    };

    ioredis.xread = async function xread(...args) {
      const result = await ioredis.sendCommand(
        ['XREAD', ...args].map((cur) =>
          typeof cur === 'number' ? `${cur}` : cur
        )
      );
      return result;
    };

    // defineCommand
    ioredis.defineCommand = async function defineLuaCommand(
      name,
      { numberOfKeys, lua }
    ) {
      const sha = await redis.scriptLoad(lua);

      ioredis[name] = async function (args: any[]) {
        const lasArgIdx = numberOfKeys ? numberOfKeys : 0;
        const keys = args.slice(0, lasArgIdx);
        let redisArgs = args
          .slice(lasArgIdx)
          .map((cur) => (typeof cur === 'number' ? `${cur}` : cur));
        if (
          redisArgs.length &&
          typeof redisArgs[redisArgs.length - 1] === 'undefined'
        ) {
          redisArgs = redisArgs.slice(0, redisArgs.length - 1);
        }
        const result = await this.evalSha(sha, {
          keys: keys,
          arguments: redisArgs,
        });
        return result;
      };

      definedCommands.set(name, ioredis[name]);
    };

    if (!ioredis.__duplicate) {
      ioredis.__duplicate = ioredis.duplicate;
      ioredis.duplicate = function duplicate(...args) {
        const newClient = this.__duplicate(...args);
        newClient.connect();
        return ioredisWrapper(newClient);
      };
    }

    if (!ioredis.__multi) {
      ioredis.__multi = ioredis.multi;
      ioredis.multi = function multi(...args) {
        const newClient = this.__multi(...args);

        newClient.__exec = newClient.exec;
        newClient.exec = async function exec(...args) {
          const results = await newClient.__exec(...args);
          return results.map((cur) => [null, cur]);
        };
        return ioredisWrapper(newClient);
      };
    }

    // Generic method proxying
    const handlers = {
      get: (target, prop, receiver) => {
        if (typeof target[prop] !== 'undefined') {
          if (typeof target[prop] === 'function') {
            return target[prop].bind(target);
          } else {
            return target[prop];
          }
        } else if (definedCommands.has(prop)) {
          return definedCommands.get(prop).bind(target);
        } else {
          return undefined;
        }
      },
    };

    return new Proxy(ioredis, handlers);
  }

  return ioredisWrapper(redis);
}
```

**Comment by slidenerd** [comment]
@manast refactor bullmq and separate it into its core and an abstraction called providers

- every provider of bullmq has a few common functions that they must implement in order for bullmq features to work
- all provider libraries need to be installed separately along with bullmq when using this library
- for example, you could have a node redis provider, an ioredis provider, a valkey provider, a keydb provider and so on
- take the library in its current form with tight coupling and try to extract ioredis functionality into a providers module just for ioredis as part of the next major version
- in the subsequent one, you could add support for other providers
- in today's date you could vibe code this abstraction with a tonne of tests that ll require code review for the most part

**`Example`**
```
npm i bullmq bullmq-ioredis-provider 
npm i bullmq bullmq-keydb-provider
npm i bullmq bullmq-node-redis-provider
```
you get the idea

**Comment by TomasTokaMrazek** [comment]
> If somebody really wants to use node-redis, one possibility would be to create a wrapper on top of node-redis that implements the IORedis specific APIs that are used by BullMQ... but I do not think this is worthy right now.

I understand the reasoning, but I think this is worth reconsidering.

Requiring users to install or wrap one Redis client while they already use another — especially the official Redis client — adds unnecessary complexity. One of the recurring problems in the JavaScript ecosystem is dependency bloat, and given the increasing number of npm supply-chain attacks, keeping dependency trees lean is becoming more important.

From that perspective, supporting both ioredis and node-redis seems like something worth at least evaluating. Other libraries have taken this approach already; for example, [redlock-universal](https://www.npmjs.com/package/redlock-universal) supports multiple Redis clients and has zero dependencies.

I realize this may not be trivial, and BullMQ may rely on APIs or behavior that are specific to ioredis. But I would much rather see support for an official client considered directly than have users build compatibility wrappers around another Redis client.

This is also a practical adoption concern: for this reason, I am hesitant to introduce BullMQ into my codebase and am more likely to choose an alternative such as RabbitMQ, where I can keep the dependency boundary clearer and reduce reliance on extra client-layer dependencies.

**Comment by fdiwa-dev** [comment]
f migrating Redis dependencies is the pain point — MiniQueue might be a simpler path. It's a REST job queue with zero Redis dependency (SQLite native). No migration, no config, just an API key.

https://github.com/fdiwa-dev/miniqueue

**Comment by avifenesh** [comment]
https://github.com/avifenesh/glide-mq
